### PR TITLE
Fixing Bug ECS 4719 Pertaining to PMPS Position State Limits Being Invalidated when State Switches to Unknown

### PR DESCRIPTION
**Pull request by NSLentz**
## Description

Added bLockBounds boolean to FB_StatePMPSEnables to lock the position limits active of the last valid goal state. Even if the axis moves out of the state without using the state mover, the position limits will remain active until a new valid goal state is selected. The bLockBounds boolean is kept true after the first valid goal state until the bEnable bit for the function block set to false. At that time, the bLockBounds boolean is reset.
Made it so that the bPowerSelf bit gets automatically set to FALSE if the FB_StatePMPSEnables block is used. This is irrespective of whether the bEnable bit for the FB_StatePMPSEnables block is true or false. If a FB_StatePMPSEnables block is called in a project the assumption is that bPowerSelf must be set to false. The alternative requires users to remember to set bPowerSelf to false and in practice this gets missed a lot.
Made it so that the PMPS position limits also prevent commands to positions outside of the allowable range instead of only looking at the current position. This will make it more difficult to command the axis out of the state fDelta range. Added formatted string error message to make it more clear when the PMPS position state limits are preventing a move.
Added a new error id to the F_MotionErrorCodeLookup. This is the error id for invalid I/O data for more than n continuous NC cycles (encoder).

## Motivation and Context

Fixed Bug https://jira.slac.stanford.edu/browse/ECS-4719 (ECS-4719) from Jira.

## How Has This Been Tested?

Tested by running full list of test cases. Added two new test cases without the bug fix incorporated. Verified the added test cases replicated the bug. Then fixed the bug and verified the test cases now passed. Needed to modify set and move helper function to make sure it initializes the fPosition value to the goal position if the axis is already at the goal position without trying to move. Added a new set and move helper function that is simpler to understand and parse.

## Where Has This Been Documented?

Documented in various code comments and also made a small update to the README to document that bPowerSelf is automatically set false when using the PMPS enables function block.

## Pre-merge checklist
- [x] Code works interactively
- [x] Test suite passes locally
- [x] Code contains descriptive comments
- [ ] Libraries are set to ``Always Newest`` version (``Library, *``)
- [x] Committed with ``pre-commit`` or ran ``pre-commit run --all-files``

### CODE REVIEW

**Comment by ZLLentz** [comment]
I think this changes enough of a central component that I also want to try it myself before hitting approve, but I'll start by reviewing the code line-by-line

File: lcls-twincat-motion/Library/POUs/Motion/PMPS/FB_StatePMPSEnables.TcPOU
Comment (ZLLentz) [inline lcls-twincat-motion/Library/POUs/Motion/PMPS/FB_StatePMPSEnables.TcPOU:146]: I'm waffling a bit in my head if it might be better to use the direct sources for these numbers rather than the EPICS status and setpoint fields. In practice these are essentially always the same but I have to wonder if there are any edge cases where either this or the raw values as reported by the NC are more correct to use.

Maybe this is best as-is due to interface flexibility (technically not requiring the NC)

But the other option would have been to use `stMotionStage.Axis.NcToPlc.ActPos` and `stMotionStage.Axis.NcToPlc.TargetPos`

Food for thought, almost didn't even post this

Edit: another thing to consider is that fPosition here is not valid until bExecute goes to TRUE, which feels like a rare edge case that could occur if someone changed code elsewhere.

File: lcls-twincat-motion/Library/POUs/Motion/PMPS/FB_StatePMPSEnables.TcPOU
Comment (ZLLentz) [inline lcls-twincat-motion/Library/POUs/Motion/PMPS/FB_StatePMPSEnables.TcPOU:151]: It's a somewhat absurd edge case, but this error isn't necessarily correct if the user wants to move positive and the PMPS restriction is "bForwardEnable = True, bBackwardEnabled = False" but instead they are hitting the positive limit to bring forth the 16#4223 error, or otherwise the general enable is False.

I'd suggest matching the directional enables with the last setpoint vs the current position.

File: lcls-twincat-motion/Library/POUs/Motion/PMPS/FB_StatePMPSEnables.TcPOU
Comment (ZLLentz) [inline lcls-twincat-motion/Library/POUs/Motion/PMPS/FB_StatePMPSEnables.TcPOU:106]: This means that we stay locked to the most recent setpoint at all times, right? I think it's a good implementation, solid and not overcomplicated.

File: lcls-twincat-motion/Library/Tests/FB_StatePMPSEnables_Test.TcPOU
Comment (ZLLentz) [inline lcls-twincat-motion/Library/Tests/FB_StatePMPSEnables_Test.TcPOU:34]: The start of file and end of file newlines are intentional to make the code more readable in review here, you don't have to include them but you probably shouldn't be removing them.

File: lcls-twincat-motion/Library/Tests/FB_StatePMPSEnables_Test.TcPOU
Comment (ZLLentz) [inline lcls-twincat-motion/Library/Tests/FB_StatePMPSEnables_Test.TcPOU:370]:
```suggestion
        'Timeout should not have occurred. Move should have been unable to complete.',
```

File: lcls-twincat-motion/_Config/NC/NC.xti
Comment (ZLLentz) [inline lcls-twincat-motion/_Config/NC/NC.xti:38]: What's with the second motor here? Is one of them unused?

**Review by ZLLentz — COMMENTED** [review]
I think this is good, I read through the code and the test carefully and pointed out the things that I had comments on.
I'm going to run this on my test plc and see how it feels as a user.

**Comment by ZLLentz** [comment]
Snazzy so far, will check a few more things in a bit
image

**Comment by ZLLentz** [comment]
Worked well in practice, can merge if it still passes tests after update branch and with comments addressed

File: lcls-twincat-motion/Library/Tests/FB_StatePMPSEnables_Test.TcPOU
Comment (NSLentz) [reply lcls-twincat-motion/Library/Tests/FB_StatePMPSEnables_Test.TcPOU:370]: The test case intentionally checks to make sure that it did timeout. E.g., the only reason it wouldn't time out is if it completed the move but completing the move is actually the failure condition. So the timeout should occur.

Comment (ZLLentz) [reply lcls-twincat-motion/Library/Tests/FB_StatePMPSEnables_Test.TcPOU:370]: Whoops, I realized this about halfway through reviewing the test but left this comment in anyway

File: lcls-twincat-motion/Library/POUs/Motion/PMPS/FB_StatePMPSEnables.TcPOU
Comment (NSLentz) [reply lcls-twincat-motion/Library/POUs/Motion/PMPS/FB_StatePMPSEnables.TcPOU:151]: I made it so that this special error message will only display if the bAllForwardEnable/bAllBackwardEnable from stMotionStage are true and they line up with the correct PMPS direction enable

File: lcls-twincat-motion/Library/Tests/FB_StatePMPSEnables_Test.TcPOU
Comment (NSLentz) [reply lcls-twincat-motion/Library/Tests/FB_StatePMPSEnables_Test.TcPOU:34]: Oh okay I didn't know that was the reasoning. I will fix.

File: lcls-twincat-motion/Library/POUs/Motion/PMPS/FB_StatePMPSEnables.TcPOU
Comment (NSLentz) [reply lcls-twincat-motion/Library/POUs/Motion/PMPS/FB_StatePMPSEnables.TcPOU:146]: I switched it to use the NC values directly. I think that makes sense

File: lcls-twincat-motion/Library/POUs/Motion/PMPS/FB_StatePMPSEnables.TcPOU
Comment (ZLLentz) [reply lcls-twincat-motion/Library/POUs/Motion/PMPS/FB_StatePMPSEnables.TcPOU:154]: Looks pretty comprehensive to me :+1:

**Review by ZLLentz — DISMISSED** [review]
I think all that's left is to click the "update branch" button that's blocking the merge button and re-run the unit tests while merged with all other changes from master

**Review by ZLLentz — APPROVED** [review]

**Comment by ZLLentz** [comment]
Admin merge to bypass CI issue after Nick re-ran and confirmed the tests

**Pull request merged**
